momaland/envs/multiwalker_stability/momultiwalker_stability.py

```diff
@@ -0,0 +1,125 @@
+"""Adapted form of the Multiwalker problem.
+
+From Gupta, J. K., Egorov, M., and Kochenderfer, M. (2017). Cooperative multi-agent control using
+deep reinforcement learning. International Conference on Autonomous Agents and Multiagent Systems
+"""
+
+from typing_extensions import override
+
+import numpy as np
+from pettingzoo.sisl.multiwalker.multiwalker import FPS
+from pettingzoo.sisl.multiwalker.multiwalker import raw_env as pz_multiwalker
+from pettingzoo.utils import wrappers
+
+from momaland.envs.multiwalker_stability.momultiwalker_stability_base import (
+    MOMultiWalkerStabilityEnv as _env,
+)
+from momaland.utils.conversions import mo_aec_to_parallel
+from momaland.utils.env import MOAECEnv
+
+
+def env(**kwargs):
+    """Returns the wrapped environment in `AEC` format.
+
+    Args:
+        **kwargs: keyword args to forward to the raw_env function.
+
+    Returns:
+        A fully wrapped AEC env.
+    """
+    env = raw_env(**kwargs)
+    env = wrappers.ClipOutOfBoundsWrapper(env)
+    return env
+
+
+def parallel_env(**kwargs):
+    """Returns the wrapped env in `parallel` format.
+
+    Args:
+        **kwargs: keyword args to forward to the raw_env function.
+
+    Returns:
+        A fully wrapped parallel env.
+    """
+    env = raw_env(**kwargs)
+    env = mo_aec_to_parallel(env)
+    return env
+
+
+def raw_env(**kwargs):
+    """Returns the environment in `AEC` format.
+
+    Args:
+        **kwargs: keyword args to forward to create the `MOMultiwalker` environment.
+
+    Returns:
+        A raw env.
+    """
+    env = MOMultiwalkerStability(**kwargs)
+    return env
+
+
+class MOMultiwalkerStability(MOAECEnv, pz_multiwalker):
+    """A sister environment to [MO-Multiwalker](https://momaland.farama.org/environments/momultiwalker), which is the MO adaptation of the [Multiwalker](https://pettingzoo.farama.org/environments/sisl/multiwalker/) environment from PettingZoo.
+
+    ## Observation Space
+    See [PettingZoo documentation](https://pettingzoo.farama.org/environments/sisl/multiwalker/#observation-space).
+
+    ## Action Space
+    The action space is a vector representing the force exerted at the 4 available joints (hips and knees), giving a continuous action space with a 4 element vector.
+    The higher bound is `1`, the lower bound is `-1`.
+
+    ## Reward Space
+    The reward space is a 2D vector where; the first value contains the sum of following rewards:
+    - Maximizing distance traveled towards the end of the level during one step. `[-0.46, 0.46]`
+    - Penalty for agent falling. `[-110, 0]`
+    - Penalty for the package falling. `[-100, 0]`
+    and the second value contains:
+    - The change in the angle of the package. `[-1.75, 0]`
```

Comment:
"A penalty based on the change of angle of the package, to avoid shaking the package."